server/models/Trust.js

```diff
@@ -42,7 +46,287 @@ class Trust {
+    }
+
+    // check if the target wallet can accept the request
+    // function below currently empty
+    // await walletModel.checkTrustRequestSentToMe(
+    //   requestType,
+    //   originatorWallet.id,
+    //   targetWallet.id,
+    // );
+
+    // create this request
+    const trustRelationship = {
+      type: TrustRelationship.getTrustTypeByRequestType(requestType),
+      request_type: requestType,
+      actor_wallet_id: actorWallet.id,
+      originator_wallet_id: originatorWallet.id,
+      target_wallet_id: targetWallet.id,
+      state: TrustRelationship.ENTITY_TRUST_STATE_TYPE.requested,
+    };
+    await this.checkDuplicateRequest({
+      walletId: originatorWallet,
+      trustRelationship,
+    });
+    const result = await this._trustRepository.create(trustRelationship);
+
+    return {
+      id: result.id,
+      actor_wallet: actorWallet.name,
+      originator_wallet: originatorWallet.name,
+      target_wallet: targetWallet.name,
+      type: result.type,
+      request_type: result.request_type,
+      state: result.state,
+      created_at: result.created_at,
+      updated_at: result.updated_at,
+      active: result.active,
+    };
+  }
+
+  // check if I (current wallet) can add a new trust like this
+  async checkDuplicateRequest({ walletId, trustRelationship }) {
+    const trustRelationships = await this.getTrustRelationships({ walletId });
+    if (
+      trustRelationship.type === TrustRelationship.ENTITY_TRUST_TYPE.send ||
+      trustRelationship.type === TrustRelationship.ENTITY_TRUST_TYPE.manage
+    ) {
+      if (
+        trustRelationships.some((e) => {
+          if (
+            (e.request_type === trustRelationship.request_type &&
+              (e.state ===
+                TrustRelationship.ENTITY_TRUST_STATE_TYPE.requested ||
+                e.state ===
+                  TrustRelationship.ENTITY_TRUST_STATE_TYPE.trusted) &&
+              e.actor_wallet_id === trustRelationship.actor_wallet_id &&
+              e.target_wallet_id === trustRelationship.target_wallet_id) ||
+            (e.request_type !== trustRelationship.request_type &&
+              (e.state ===
+                TrustRelationship.ENTITY_TRUST_STATE_TYPE.requested ||
+                e.state ===
+                  TrustRelationship.ENTITY_TRUST_STATE_TYPE.trusted) &&
+              e.actor_wallet_id === trustRelationship.target_wallet_id &&
+              e.target_wallet_id === trustRelationship.actor_wallet_id)
+          ) {
+            return true;
+          }
+          return false;
+        })
+      ) {
+        log.debug('Has duplicated trust');
+        throw new HttpError(
+          403,
+          'The trust relationship has been requested or trusted',
+        );
+      }
+    } else {
+      throw HttpError(500, 'Not supported type');
+    }
+    log.debug('Has no duplicated trust');
+  }
+
+  async checkManageCircle({ walletId, trustRelationship }) {
```

Comment:
I reason I didn't use SQL to do the job is because, yes we can do all the business through SQL, but if we bring all business into repository, then, the model is missing, and all the responsibility goes to SQL

Reply:
Yeah, I understand. I think that is fine. I also tried to maintain that where I can